gax/pagination.h

```diff
@@ -51,25 +74,185 @@ class PageResult {
     }
     bool operator!=(iterator const& rhs) const { return !(*this == rhs); }
 
+   private:
+    friend PageResult;
     iterator(FieldIterator current) : current_(std::move(current)) {}
 
     FieldIterator current_;
   };
-  // Note: Since the constructor will eventually be private (with friend
-  // access), we don't have to define all the variants for references.
-  // We just want to take ownership of a page.
-  PageResult(PageType&& raw_page, ElementAccessor accessor)
-      : raw_page_(std::move(raw_page)), accessor_(accessor) {}
 
-  iterator begin() { return accessor_(raw_page_)->begin(); }
-  iterator end() { return accessor_(raw_page_)->end(); }
+  PageResult(PageType const& raw_page) : raw_page_(raw_page) {}
+  PageResult(PageType&& raw_page) : raw_page_(std::move(raw_page)) {}
 
+  iterator begin() { return ElementAccessor{}(raw_page_)->begin(); }
```

Comment:
Why construct the ElementAccessor over and over again? Can't we retain the accessor_ field and use that? Or does that add to the memory footprint and not actually result in any improvement?

Reply:
The accessor has no fields and a trivial constructor, and the overload to operator() should hopefully be small enough to be inlined.
Saving an instance of ElementAccessor requires at least 1 byte of space; I haven't run any benchmarks, but it should never be more expensive to recreate the ElementAccessor whenever it's used.